### ISSUE TITLE
New onCompleted callback in Query is getting called when an error happen

### ISSUE THREAD

**Issue by sandrocsimas** [heftapp/graphql_codegen#257]
I have the following query:
```
useQuery$GetMe(Options$Query$GetMe(
    onComplete: (_, __) {
        WidgetsBinding.instance.addPostFrameCallback((_) {
            context.pushReplacement(const Home());
        });
    },
    onError: (_) {
        WidgetsBinding.instance.addPostFrameCallback((_) {
            context.pushReplacement(const SignIn());
        });
    },
));
```
This query generates an error when Authorization header is not present.
```
{
  "data": {},
  "errors": [
    {
      "message": "Access denied! You need to be authorized to perform this action!",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "me"
      ],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "stacktrace": [
          "Error: Access denied! You need to be authorized to perform this action!",
          "    at /home/sandro/workspace/closic-core/node_modules/type-graphql/dist/helpers/auth-middleware.js:20:44",
          "    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)",
          "    at async dispatchHandler (/home/sandro/workspace/closic-core/node_modules/type-graphql/dist/resolvers/helpers.js:83:24)"
        ]
      }
    }
  ]
}
```
The onCompleted callback is called instead of onError callback.
Here is the stacktrace:
```
E/flutter ( 5458): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: type 'Null' is not a subtype of type 'Map<String, dynamic>' in type cast
E/flutter ( 5458): #0      new Query$GetMe.fromJson (package:closic/graphql/generated/user.graphql.dart:18:41)
E/flutter ( 5458): #1      _parserFn$Query$GetMe (package:closic/graphql/generated/user.graphql.dart:303:17)
E/flutter ( 5458): #2      new Options$Query$GetMe.<anonymous closure> (package:closic/graphql/generated/user.graphql.dart:334:21)
E/flutter ( 5458): #3      QueryCallbackHandler.onCompleted.<anonymous closure> (package:graphql/src/core/query_options.dart:369:37)
E/flutter ( 5458): #4      ObservableQuery._applyCallbacks (package:graphql/src/core/observable_query.dart:318:21)
E/flutter ( 5458): #5      ObservableQuery.addResult (package:graphql/src/core/observable_query.dart:274:7)
E/flutter ( 5458): #6      QueryManager.addQueryResult (package:graphql/src/core/query_manager.dart:409:23)
E/flutter ( 5458): #7      QueryManager._resolveQueryOnNetwork (package:graphql/src/core/query_manager.dart:284:7)
```

**Comment by sandrocsimas** [comment]
The error happens before my onCompleted function, as you can see in the stacktrace.

**Comment by budde377** [comment]
I think the callback is called at the right time, but it should not error.

**Comment by budde377** [comment]
A fix is being released as `0.12.0-beta.15` now. Please make sure that you upgrade `graphql` to `5.2.0-beta.1` or later.

**Comment by sandrocsimas** [comment]
Great job, many thanks! Let me try again. 👏🏽👏🏽

**Comment by sandrocsimas** [comment]
Hi @budde377, the same error is happening. I think it is because of this:
`return options.onComplete!(result.data ?? {});`

**Comment by budde377** [comment]
Where are you seeing this?

**Comment by sandrocsimas** [comment]
```
  OnData<TParsed>? get onCompleted {
    if (options.onComplete != null) {
      return (QueryResult? result) {
        if (!result!.isLoading && !result.isOptimistic) {
          return options.onComplete!(result.data ?? {});
        }
      };
    }
    return null;
  }
```
This code is in `query_options.dart` from `graphql-5.1.3`

Code gen is checking null, not an empty map.

**Comment by sandrocsimas** [comment]
As the data is an empty map, the following cast fails because `l$me` is null.

```
factory Query$GetMe.fromJson(Map<String, dynamic> json) {
    final l$me = json['me'];
    final l$$__typename = json['__typename'];
    return Query$GetMe(
      me: Query$GetMe$me.fromJson((l$me as Map<String, dynamic>)),
      $__typename: (l$$__typename as String),
    );
  }
```

**Comment by sandrocsimas** [comment]
In `mutation_options.dart` (graphql-5.1.3) the null value is sent as parameter instead of an empty map:
```
  OnData<TParsed>? get onCompleted {
    if (options.onCompleted != null) {
      return (QueryResult? result) {
        if (!result!.isLoading && !result.isOptimistic) {
          return options.onCompleted!(result.data);
        }
      };
    }
    return null;
  }
```
That's why it works for mutation.

**Comment by sandrocsimas** [comment]
I just figured out that you fixed that in graphql_flutter beta version:
https://github.com/zino-hofmann/graphql-flutter/commit/aceff150bf8c676207f16aa07f9622dfc7e56e9d
